sklearn/metrics/_plot/confusion_matrix.py

```diff
@@ -83,22 +83,30 @@ def plot(self, include_values=True, cmap='viridis',
         n_classes = cm.shape[0]
         self.im_ = ax.imshow(cm, interpolation='nearest', cmap=cmap)
         self.text_ = None
-
         cmap_min, cmap_max = self.im_.cmap(0), self.im_.cmap(256)
 
         if include_values:
             self.text_ = np.empty_like(cm, dtype=object)
-            if values_format is None:
-                values_format = '.2g'
 
             # print text with appropriate color depending on background
             thresh = (cm.max() + cm.min()) / 2.0
+
             for i, j in product(range(n_classes), range(n_classes)):
                 color = cmap_max if cm[i, j] < thresh else cmap_min
-                self.text_[i, j] = ax.text(j, i,
-                                           format(cm[i, j], values_format),
-                                           ha="center", va="center",
-                                           color=color)
+
+                fmt = values_format
+                if values_format is None:
```

Comment:
I think trying `.2g` and then trying `d` if applicable, is slightly clearer

This suggestion is to address https://github.com/scikit-learn/scikit-learn/pull/16159#discussion_r370265170

```py
if values_format is None:
    text_cm = format(cm[i, j], '.2g')
    if cm.dtype.kind != 'f':
        text_d = format(cm[i, j], 'd')
        if len(text_d) < len(text_cm):
            text_cm = text_d
else:
    text_cm = format(cm[i, j], values_format)
```

The test would need to updated to reflect this behavior.